spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -18,193 +18,79 @@
-    }
-
-    Row currentRow = currentRow();
-
-    if (currentRow.getString(changeTypeIndex).equals(DELETE) && rowIterator.hasNext()) {
-      Row nextRow = rowIterator.next();
-      cachedRow = nextRow;
-
-      if (isUpdateOrCarryoverRecord(currentRow, nextRow)) {
-        if (isCarryoverRecord(currentRow, nextRow)) {
-          // set carry-over rows to null for filtering out later
-          currentRow = null;
-          cachedRow = null;
-        } else {
-          currentRow = modify(currentRow, changeTypeIndex, UPDATE_BEFORE);
-          cachedRow = modify(nextRow, changeTypeIndex, UPDATE_AFTER);
-        }
-      }
-    }
-
-    return currentRow;
-  }
-
-  private Row modify(Row row, int valueIndex, Object value) {
-    if (row instanceof GenericRow) {
-      GenericRow genericRow = (GenericRow) row;
-      genericRow.values()[valueIndex] = value;
-      return genericRow;
-    } else {
-      Object[] values = new Object[row.size()];
-      for (int index = 0; index < row.size(); index++) {
-        values[index] = row.get(index);
-      }
-      values[valueIndex] = value;
-      return RowFactory.create(values);
-    }
-  }
-
-  private int[] generateIndicesForIdentifySameRow(int columnSize) {
-    int[] indices = new int[columnSize - 1];
-    for (int i = 0; i < indices.length; i++) {
-      if (i < changeTypeIndex) {
-        indices[i] = i;
-      } else {
-        indices[i] = i + 1;
-      }
-    }
-    return indices;
-  }
-
-  private boolean isCarryoverRecord(Row currentRow, Row nextRow) {
-    for (int idx : indicesForIdentifySameRow) {
-      if (!isColumnSame(currentRow, nextRow, idx)) {
-        return false;
-      }
-    }
-
-    return true;
-  }
-
-  private boolean cachedUpdateRecord(Row cachedRecord) {
-    return cachedRecord != null
-        && !cachedRecord.getString(changeTypeIndex).equals(DELETE)
-        && !cachedRecord.getString(changeTypeIndex).equals(INSERT);
-  }
-
-  private Row currentRow() {
-    if (cachedRow != null) {
-      Row row = cachedRow;
-      cachedRow = null;
-      return row;
-    } else {
-      return rowIterator.next();
-    }
+  /**
+   * Creates an iterator that removes carry-over rows from a changelog table.
+   *
+   * @param rowIterator the iterator of rows from a changelog table
+   * @param rowType the schema of the rows
+   * @return a new iterator instance
+   */
+  public static Iterator<Row> createCarryoverRemoveIterator(
+      Iterator<Row> rowIterator, StructType rowType) {
+    CarryoverRemoveIterator changelogIterator = new CarryoverRemoveIterator(rowIterator, rowType);
+    return Iterators.filter(changelogIterator, Objects::nonNull);
   }
 
-  private boolean isUpdateOrCarryoverRecord(Row currentRow, Row nextRow) {
-    return sameLogicalRow(currentRow, nextRow)
-        && currentRow.getString(changeTypeIndex).equals(DELETE)
-        && nextRow.getString(changeTypeIndex).equals(INSERT);
+  protected boolean isColumnSame(Row currentRow, Row nextRow, int idx) {
```

Comment:
optional: It seems the result of this method is always negated. I am wondering whether we want to switch the logic around.

```
protected boolean isDifferentValue(Row currentRow, Row nextRow, int idx) {
  return !Objects.equals(nextRow.get(idx), currentRow.get(idx));
}
```